File: app/src/main/java/info/firozansari/android_intent_example/utils/ItemAdapter.java
package info.firozansari.android_intent_example.utils;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import info.firozansari.android_intent_example.DemoItem;
import info.firozansari.android_intent_example.MainFragment;
import info.firozansari.android_intent_example.R;

public class ItemAdapter extends RecyclerView.Adapter<ItemAdapter.ViewHolder>{

    private List<DemoItem> itemList = new ArrayList<>();
    private Context context;
    private MainFragment.OnFragmentInteractionListener listener;


    public ItemAdapter(Context context, MainFragment.OnFragmentInteractionListener listener) {
        this.context = context;
        this.listener = listener;
    }


    @Override
    public int getItemCount() {
        if(itemList != null) {
            if (itemList.size() == 0) {
                return 0;
            } else {
                return itemList.size();
            }
        }
        return 0;
    }

    @Override
    public ItemAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        return new ItemAdapter.ViewHolder(LayoutInflater.from(context).inflate(R.layout.list_item, parent, false));
    }

    @Override
    public void onBindViewHolder(ItemAdapter.ViewHolder holder, int position) {
        if (itemList.size() > 0) {

            DemoItem demoItem = itemList.get(position);

            if(demoItem != null) {
                holder.issueTitle.setText(demoItem.getDescription());
            }
        }

    }


    // Provide a reference to the views for each data item
    // Complex data items may need more than one view per item, and
    // you provide access to all the views for a data item in a view holder
    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        public TextView issueTitle;

        public ViewHolder(View v) {
            super(v);
            issueTitle = (TextView) v.findViewById(R.id.item_tv);
            issueTitle.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            DemoItem item = getItem(getAdapterPosition());
            listener.onFragmentInteraction(item.getIntent());
        }
    }

    private DemoItem getItem(int adapterPosition) {

        if (adapterPosition >= 0) {
            return itemList.get(adapterPosition);
        }

        return null;
    }


    public void swapList(List<DemoItem> demoItemList) {
        if(demoItemList != null) {
            this.itemList.clear();
            this.itemList.addAll(demoItemList);
            notifyDataSetChanged();
        }

    }

}

File: app/src/main/java/info/firozansari/android_intent_example/utils/ItemAdapter.kt
package info.firozansari.android_intent_example.utils

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import info.firozansari.android_intent_example.DemoItem
import info.firozansari.android_intent_example.MainFragment.OnFragmentInteractionListener
import info.firozansari.android_intent_example.R

class ItemAdapter(
    private val context: Context,
    private val listener: OnFragmentInteractionListener
) : RecyclerView.Adapter<ItemAdapter.ViewHolder>() {
    private val itemList: MutableList<DemoItem>? = ArrayList()


    override fun getItemCount(): Int {
        if (itemList != null) {
            return if (itemList.size == 0) {
                0
            } else {
                itemList.size
            }
        }
        return 0
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            LayoutInflater.from(
                context
            ).inflate(R.layout.list_item, parent, false)
        )
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        if (itemList!!.size > 0) {
            val demoItem = itemList[position]

            if (demoItem != null) {
                holder.issueTitle.text = demoItem.description
            }
        }
    }


    // Provide a reference to the views for each data item
    // Complex data items may need more than one view per item, and
    // you provide access to all the views for a data item in a view holder
    inner class ViewHolder(v: View) : RecyclerView.ViewHolder(v), View.OnClickListener {
        var issueTitle: TextView = v.findViewById<View>(R.id.item_tv) as TextView

        init {
            issueTitle.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            val item = getItem(adapterPosition)
            listener.onFragmentInteraction(item!!.intent)
        }
    }

    private fun getItem(adapterPosition: Int): DemoItem? {
        if (adapterPosition >= 0) {
            return itemList!![adapterPosition]
        }

        return null
    }


    fun swapList(demoItemList: List<DemoItem>?) {
        if (demoItemList != null) {
            itemList!!.clear()
            itemList.addAll(demoItemList)
            notifyDataSetChanged()
        }
    }
}
